# U.S. Medical Insurance Costs
In this project, I will focus on validating whether the cost for those who smoke is higher than for those who do not, and whether smoking is more common among men or women. I know there are libraries like Pandas that can make the analysis easier, but I want to keep it simple in order to practice the skills I’ve learned so far.






### Import libraries
To start we have to import all the libraries that we may want to use for this project.

In [14]:
import csv
from collections import Counter

### Loading dataset and exploring data
In this section, we will load our medical insurance dataset and take a look at the data.

In [15]:
# Code section to read insurance.csv file
file_path = "insurance.csv"

with open(file_path, newline="") as insurance_file:
    data_reader = csv.DictReader(insurance_file)
    data = list(data_reader)
# Print first five rows
print(data[:5])

[{'age': '19', 'sex': 'female', 'bmi': '27.9', 'children': '0', 'smoker': 'yes', 'region': 'southwest', 'charges': '16884.924'}, {'age': '18', 'sex': 'male', 'bmi': '33.77', 'children': '1', 'smoker': 'no', 'region': 'southeast', 'charges': '1725.5523'}, {'age': '28', 'sex': 'male', 'bmi': '33', 'children': '3', 'smoker': 'no', 'region': 'southeast', 'charges': '4449.462'}, {'age': '33', 'sex': 'male', 'bmi': '22.705', 'children': '0', 'smoker': 'no', 'region': 'northwest', 'charges': '21984.47061'}, {'age': '32', 'sex': 'male', 'bmi': '28.88', 'children': '0', 'smoker': 'no', 'region': 'northwest', 'charges': '3866.8552'}]


In [16]:
# As can be observed above, the numeric variables are in text format, so we will
# create a function to convert the values of the numeric variables into integers
# or decimal numbers
def normalize_row(row):
    normalized = {
        "age": int(row["age"]),
        "sex": row["sex"],
        "bmi": round(float(row["bmi"]), 2),
        "children": int(row["children"]),
        "smoker": row["smoker"],
        "region": row["region"],
        "charges": round(float(row["charges"]), 2)
    }
    return normalized

In [17]:
insurance_costs = [normalize_row(row) for row in data]
print(insurance_costs[:5])
# At this point, our data is ready to be used, and we will create some functions
# that will help us avoid writing too much code

[{'age': 19, 'sex': 'female', 'bmi': 27.9, 'children': 0, 'smoker': 'yes', 'region': 'southwest', 'charges': 16884.92}, {'age': 18, 'sex': 'male', 'bmi': 33.77, 'children': 1, 'smoker': 'no', 'region': 'southeast', 'charges': 1725.55}, {'age': 28, 'sex': 'male', 'bmi': 33.0, 'children': 3, 'smoker': 'no', 'region': 'southeast', 'charges': 4449.46}, {'age': 33, 'sex': 'male', 'bmi': 22.7, 'children': 0, 'smoker': 'no', 'region': 'northwest', 'charges': 21984.47}, {'age': 32, 'sex': 'male', 'bmi': 28.88, 'children': 0, 'smoker': 'no', 'region': 'northwest', 'charges': 3866.86}]


### Functions
Here I will create functions to perform category counts, calculate totals, and compute the mean.

In [18]:
# Function to calculate the average value of a numeric key across a list of dictionaries
def calculate_mean(data, col_name):
    total = 0
    num_rows = len(data)
    for row in data:
        total += row[col_name]
    return round(total / num_rows, 2)

# Function to calculate the total value of a numeric key across a list of dictionaries
def calculate_total(data, col_name):
    total = 0
    for row in data:
        total += row[col_name]
    return total

# Function to count the occurrences of a category key across a list of dictionaries
def count_category(data, col_name):
    categories = list(row[col_name] for row in data)
    return dict(Counter(categories))

### Analisys

In [19]:
# Total Observations
num_rows = len(insurance_costs)
print(f"Count of total observations: {num_rows}")

# Total Smokes
num_smokers = count_category(insurance_costs, "smoker")
print(f"Smokers: {num_smokers}")

# Total by sex 
num_genders = count_category(insurance_costs, "sex")
print(f"Genders: {num_genders}")

Count of total observations: 1338
Smokers: {'yes': 274, 'no': 1064}
Genders: {'female': 662, 'male': 676}


In [20]:
# Create a list of smokers and non-smokers
smokers = list(row for row in insurance_costs if row["smoker"] == "yes")
no_smokers = list(row for row in insurance_costs if row["smoker"] == "no")

# Compute the mean of charges for smokers and non-smokers
avg_smokers_charges = calculate_mean(smokers, "charges")
avg_no_smokers_charges = calculate_mean(no_smokers, "charges")

print("Avg Charges Smokers: {}".format(avg_smokers_charges))
print("Avg Charges Non-Smokers: {}".format(avg_no_smokers_charges))
print("Difference: {}".format(avg_smokers_charges - avg_no_smokers_charges))


Avg Charges Smokers: 32050.23
Avg Charges Non-Smokers: 8434.27
Difference: 23615.96


In [21]:
# Total smokers by sex
num_smokers_by_sex = count_category(smokers, "sex")
print(f"Total Smokers: {len(smokers)}")
print(f"Smokers by Sex: {num_smokers_by_sex}")

Total Smokers: 274
Smokers by Sex: {'female': 115, 'male': 159}


In [22]:
# Create a list of male and female smokers 
male_smokers = list(row for row in smokers if row["sex"] == "male")
female_smokers = list(row for row in smokers if row["sex"] == "female")

# Calculate de mean of male_smokers and female_smokers
avg_male_smokers = calculate_mean(male_smokers, "charges")
avg_female_smokers = calculate_mean(female_smokers, "charges")

print(f"Avg Charges Male Smoker: {avg_male_smokers}")
print(f"Avg Charges Female Smoker: {avg_female_smokers}")
print(f"Difference: {avg_male_smokers - avg_female_smokers}")

Avg Charges Male Smoker: 33042.01
Avg Charges Female Smoker: 30679.0
Difference: 2363.010000000002


### Conclutions
* Smokers pay about $23,615.96 more on average than Non-Smokers
  
* People who smoke represent 20.47% of the medical insurance dataset. The average difference in charges between male and female smokers is $2,363.01, which is much lower than the additional costs smokers in general tend to pay.